src/controller/java/OTAProviderDelegateBridge.cpp

```diff
@@ -0,0 +1,375 @@
+        // timeout timer while preparing for transfer if any failure occurs afterwards.
+        mBdxOTASender->ResetState();
+        sendOTAQueryFailure(static_cast<uint8_t>(OTAQueryStatus::kNotAvailable));
+        return;
+    }
+
+    char uriBuffer[kMaxBDXURILen];
+    MutableCharSpan uri(uriBuffer);
+    err = bdx::MakeURI(ourNodeId.GetNodeId(), jniFilePath.c_str() != nullptr ? jniFilePath.charSpan() : CharSpan(), uri);
+    if (CHIP_NO_ERROR != err)
+    {
+        LogErrorOnFailure(err);
+        commandObj->AddStatus(cachedCommandPath, StatusIB(err).mStatus);
+        mBdxOTASender->ResetState();
+        return;
+    }
+    response.imageURI.SetValue(uri);
+    commandObj->AddResponse(cachedCommandPath, response);
+}
+
+void OTAProviderDelegateBridge::HandleApplyUpdateRequest(CommandHandler * commandObj, const ConcreteCommandPath & commandPath,
+                                                         const ApplyUpdateRequest::DecodableType & commandData)
+{
+    assertChipStackLockedByCurrentThread();
+
+    VerifyOrReturn(mOtaProviderDelegate != nullptr,
+                   commandObj->AddStatus(commandPath, Protocols::InteractionModel::Status::Failure));
+
+    NodeId nodeId = commandObj->GetSubjectDescriptor().subject;
+
+    ConcreteCommandPath cachedCommandPath(commandPath.mEndpointId, commandPath.mClusterId, commandPath.mCommandId);
+
+    JNIEnv * env = JniReferences::GetInstance().GetEnvForCurrentThread();
+
+    jmethodID handleApplyUpdateRequestMethod;
+    CHIP_ERROR err = JniReferences::GetInstance().FindMethod(env, mOtaProviderDelegate, "handleApplyUpdateRequest",
+                                                             "(JJ)Lchip/devicecontroller/OTAProviderDelegate$ApplyUpdateResponse;",
+                                                             &handleApplyUpdateRequestMethod);
+    VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(Controller, "Could not find handleApplyUpdateRequest method"));
+
+    jobject jResponse = env->CallObjectMethod(mOtaProviderDelegate, handleApplyUpdateRequestMethod, static_cast<jlong>(nodeId),
+                                              static_cast<jlong>(commandData.newVersion));
+    if (env->ExceptionCheck())
+    {
+        ChipLogError(Support, "Exception in call java method");
+        env->ExceptionDescribe();
+        env->ExceptionClear();
+        return;
+    }
+
+    jmethodID getActionMethod;
+    err = JniReferences::GetInstance().FindMethod(env, jResponse, "getAction", "()I", &getActionMethod);
+    VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(Controller, "Could not find getAction method"));
+
+    jmethodID getDelayedActionTimeMethod;
+    err = JniReferences::GetInstance().FindMethod(env, jResponse, "getDelayedActionTime", "()J", &getDelayedActionTimeMethod);
+    VerifyOrReturn(err == CHIP_NO_ERROR, ChipLogError(Controller, "Could not find getDelayedActionTime method"));
+
+    jint jAction             = env->CallIntMethod(jResponse, getActionMethod);
```

Comment:
both jAction and jDelayedActionTime are leaking local ref

Reply:
Add "chip::JniLocalReferenceManager manager(env);" in this function